### DISCUSSION TITLE
Design survey: Transform architecture

### DISCUSSION THREAD

**Discussion by alice-i-cecile** [bevyengine/bevy#4213] [Ideas]
This discussion is intended to collect issues and proposals in a central place, in order to provide an overview of the rather complex set of concerns that we should slowly try to untangle.

If you have additional challenges or constraints, please post them and I'll update this list. If you have design proposals, please post them here!

# Current challenges

1. There is severe code duplication between `Transform` and `GlobalTransform`.
    1. Discussion in #4168.
2. The "transform propagation system" for inheriting transforms cannot be readily extended because `Transform` is not provided as an associated type in any form for `GlobalTransform` or vice versa.
     1. Users cannot opt-out of inheriting `Transform`, or any of its constituents.
4. Code cannot work with "either `Transform` or `GlobalTransform`" without a custom enum wrapper.
5. Users commonly struggle to understand the distinction and relation between `Transform` and `GlobalTransform`, and which one they should be modifying or reading for their various use cases.
    1. This has improved with better docs, and could be improved further. 
    2. Meddling with `GlobalTransform` almost never has the desired results, and meddling with the transform values of UI elements, while extremely tempting, tends to lead to either no effect or very strange behavior.
6. Non-uniform scaling is of questionable value, and is likely to have performance costs.
    1. #4207
    2. #1755
7. Transforms cannot be easily extended to support other types with other behaviors, data representations or improved ergonomics.
    1. `f64` (#1680)
    2. Discretized coordinates: both grid-like and fixed point (#1680)
    4. UI: https://github.com/bevyengine/rfcs/pull/53, #1257
    5. 2D: #4156, #2548
    6. Non-Euclidean transforms: #4207
9. The ergonomics (and performance) of working with the full `Transform` when you only need to mutate a single constituent (such as the translation) is debatable.
    1. Inspired the contentious #2252, which was ultimately rejected for being confusing to users.

# Constraints

1. Regardless of their coordinate system, all `Transform` types must ultimately be renderable.
2. Performance is very important, at least in some applications.
3. Avoiding heavy abuse of generics and formal mathematical constructs in public APIs (see nalgebra) is important to ergonomics and accessibility.
4. Different types of transforms will commonly be blended within the same game, depending on the object type.

**Comment by alice-i-cecile** [comment]
# Use macros to reduce `Transform` and `GlobalTransform` code duplication

This reduces the immediate pain, but significantly increases developer complexity. Macros don't always play nicely with tooling, and are intimidating for many new developers.

**Comment by alice-i-cecile** [comment]
# Common `RenderTransform` representation

1. Store a `Mat4` or `Affine3A` in a `RenderTransform` component.
2. Allow users to use their own transformlike components by supplying them to the rendering plugin, but enforce `Into<RenderTransform>`.
3. Use the `RenderTransform` to handle rendering.

First raised in #4207 by @finegeometer.

**Reply by james7132** [reply]
If the goal is to support non-Euclidean geometries, Mat4/Afffine3A might not be enough, since not all projections might fit in such a representation or be efficiently stored in it either. Depth calculations for transparencies and shadow projections would also likely need to be provided via trait or function pointer.

Definitely would not like to make PbrPlugin or RenderPlugin generic over the space it operates in.

**Reply by CptPotato** [reply]
I have no experience with non euclidian geometry so I can't really comment on that part.

But other than that, I think `Affine3A` is sufficient for pretty much all general use cases where you "place objects in 3d space". The one thing it cannot encode is perspective projection, so cameras will still need to use `Mat4` for at least their projection matrix.

**Reply by finegeometer** [reply]
For some context, here's why I suggested this.

- Eventually, we will want multiple `Transform` types.
    - We may want dedicated `Transform2D` or `TransformUI` types; see the below discussions.
    - We may want an `f64`-based `Transform` type, for when more precision is needed; see #1680.
    - Personally, I'm interested in non-Euclidean geometry, so I'd want to define a `SphericalTransform` and `HyperbolicTransform`.
- User code will want `GlobalTransform`'s representation to match that of `Transform`.
    - In #4207, I used the example of setting a bullet's initial position to be inside a gun, when the gun is deep in a hierarchy. The code would set the bullet's `Transform` to match the gun's `GlobalTransform`.
- Rendering code wants there to be a single `GlobalTransform` type. There are all sorts of rendering systems that refer to it; systems we'd rather not be generic.

These three facts don't work well together. `GlobalTransform` is required by user code to be several types, but required by render code to be one type.

My current best idea for solving this is to have multiple `GlobalTransform` types, sharing the representation of their corresponding `Transform`s, and one `RenderTransform` type, represented as an `Affine3A` or `Mat4`. Then user code would work with the former, render code would work with the latter, and there would be some system connecting them.

> If the goal is to support non-Euclidean geometries, `Mat4`/`Afffine3A` might not be enough, since not all projections might fit in such a representation or be efficiently stored in it either.

OpenGL makes it easy to work with geometries whose `Transform`s are nicely represented as `Mat4`s, such as spherical or hyperbolic geometry. In #4207, those are the only geometries I suggest supporting, for that reason. So this isn't a problem, because I was only asking about those geometries for which it isn't a problem!

> I have no experience with non-Euclidean geometry so I can't really comment on that part.
>
> But other than that, I think `Affine3A` is sufficient for pretty much all general use cases where you "place objects in 3d space". The one thing it cannot encode is perspective projection, so cameras will still need to use `Mat4` for at least their projection matrix.

As you say, `Affine3A` is sufficient for Euclidean geometry. `Mat4` supports spherical and hyperbolic geometry as well.

We should probably start with `Affine3A`, and come back to the non-Euclidean case after we've gotten everything else worked out.

**Reply by james7132** [reply]
> OpenGL makes it easy to work with geometries whose Transforms are nicely represented as Mat4s, such as spherical or hyperbolic geometry. In https://github.com/bevyengine/bevy/discussions/4207, those are the only geometries I suggest supporting, for that reason. So this isn't a problem, because I was only asking about those geometries for which it isn't a problem!

I replied in that thread how one might be able to achieve the goals they want without forking any of the rendering crates. The sole exceptions to this are mid-pipeline depth sorting for transparent objects and shadow projections, which I'm not sure we can achieve without making the render plugins generic over the space they operate over or exposing function pointer parameters for handling this.

**Reply by finegeometer** [reply]
Now that I understand the rendering architecture a bit better, I think I have a clearer understanding of the design space.

Users work with `Transform`. Rendering works with `Mat4`. So somewhere along the line, a conversion has to happen.

When should this conversion happen? There are four choices:
1. Do the conversion in the app world.
2. Do the conversion during the `Extract` step.
3. Do the conversion in the render world.
4. Some combination of the above.

The status quo is option 4 — option 2 for meshes; option 3 for cameras and point lights.
In my understanding, this proposal is option 1, but I'm not confident I'm interpreting the proposal in the same way as @alice-i-cecile.

In terms of ease-of-replacing-`Transform`, I think option 1 is better than option 2, which is better than option 3. But I could be missing something, or there could be other considerations which lead to a different preference order.

**Reply by james7132** [reply]
Option 1 might be workable if and only if we can justify the cost in blocking all other systems reliant on a GlobalTransform equivalent. As we're already doing composition, adding the cost of a matrix multiply alongside conversion into a Mat4 may increase transform propagation time significantly. Main benefit here is that the Mat4/Affine3A representation is available to everything, not just rendering. Could also regain some time back by leveraging SIMD acceleration. This propagation only blocks the systems dependent on it, not the entire ECS schedule.

Option 2 shifts that to a exclusive stage that blocks both normal game simulation and rendering. IMO this is not a viable option in any case.

Option 3 is, performance wise, the most appealing, as it keeps a minimal representation in the app world, has a simple clone-to-extract process, and conversion into a GPU friendly format does not block game simulation in any way with pipelining. However, customizing it's calculation will require either generics or function pointers of some kind.

**Reply by superdump** [reply]
I’m not sure why you say Option 3 is the most appealing. As I recall, transform propagation is faster when using Mat4s and fastest when using Affine3A. Overall performance with serial rendering depends on end to end time taken to process a frame. Pipelining makes it more about minimising the maximum processing time of simulation and rendering, because the one that takes longer is the bottleneck. If rendering is the bottleneck then offloading work from the render app to the main app could improve frame rate.

**Reply by superdump** [reply]
I feel like Transform should use a separate representation and be suitable for editing. For that it should have lossless setting/getting of values. @cart has said they want this multiple times.

I feel that the internal representation of GlobalTransform should rather be optimal for internal use as it has associated performance costs.

That said, as the ‘spawn something at the global position of another thing, orientated correctly’ requires supporting some use cases to get information from the global transform so we need to make sure those are also supported.

We should be aware of what other engines support as expectations will be that we can at least support the same functionality. To my understanding, Unity uses a separate representation for local transforms and affine matrix representation for global.

I feel that supporting non-affine global transforms does not have to be a requirement of the core of bevy and rather a plugin could provide support for spherical and hyperbolic geometries and rendering. I feel that if they can be supported without significantly negatively affecting performance for affine geometries then we definitely can but it seems that only sorting affine geometries is the external expectation and doing so does provide significant performance benefits.

**Comment by alice-i-cecile** [comment]
# Dedicated 2D Transform type

**Reply by james7132** [reply]
```rust
pub struct Transform2D {
  pub translation: Vec2,
  pub mat2: Mat2,
  pub depth: f32,
}
```

A few notes:

 - This would effectively be `glam::Affine2` with an additional depth parameter. 
 - Scale and rotation are encoded in the `Mat2`.
 - 2D transforms are where non-uniform scale is definitely desirable. See: https://www.youtube.com/watch?v=haa7n3UGyDc
 - One open question is whether `depth` should be inheritable/composable. IMO depth in 2D should be treated as a near static layer, so a child should typically have a depth that is independent of it's parents.
 - This implementation has zero SIMD acceleration.

**Reply by CptPotato** [reply]
An alternative representation would be this:

```rust
pub struct Transform2D {
  pub translation: Vec2,
  pub scale: Vec2,
  pub rotation: f32,
  pub depth: f32,
}
```

The main difference is that it is more in line with the current `Transform` layout. And by keeping scale and rotation separate, relative rotation can be applied without having it affect scale over time (and vice versa).

**Reply by james7132** [reply]
One potential problem is that any rotation using a 1D rotation requires trig function usage (sin/cos) when propagating. Unless we are hyper aggressive in how optimize that, this may have a heavy performance penalty. Some common methods for simplifying this would be to include a big LUT and use linear interpolation to approximate, but even this is not optimal.

An alternative, to steal the 3D quaternion idea, is to use a dedicated complex number to encode rotation instead. This would be closed under any rotation operation. You won't get any unbounded f32 values provided stick to values on the unit-circle, and it'd be near trivial to make a 2D rotation matrix out of it.

**Reply by CptPotato** [reply]
Good point. I mainly thought about it from an UX standpoint for when you want to read or adjust the scale/rotation. If it's a matrix you have to decompose the `Mat2`, adjust rot/scale and then create the `Mat2` again.

I also considered using a `Vec3` for the rotation: rotation `angle` + pre computed `cos` and `sin` (updated on assignment).
Though, now that I think about it, this would basically be your proposed solution plus having the user add an additional component that holds the raw rotation & scale separately for the entities where it is required.

**Reply by superdump** [reply]
@cart had been pretty clear previously about wanting to have a separate representation for Transform for local, parent-relative transforms for editing so that what the user sees in the editor is what the user gets after setting a value. That suggests that for user-facing Transform, the Affine2 or otherwise matrix representation is out.

**Comment by alice-i-cecile** [comment]
# Dedicated UI Transform type

Goals:

1. Focused on rectangular layouts.
2. No rotation
3. 2D only
4. Some sort of z-layering solution

Currently, we're storing Transform, GlobalTransform, Size, and CalculatedSize, but only using the latter two. This is both inefficient and misleading.

**Reply by james7132** [reply]
```rust
pub struct RectTransform {
  pub position: Vec2,
  pub size: Size,
}

pub struct GlobalRectTransform {
  pub position: Vec2,
  pub size: Size,
  pub depth: u32,
}
```

Some notes:
 
 - Position and size are driven by the layouting algorithm for UI.
 - Position is in the top-left corner of the rect relative to the position of the entity's parent. Propagation is simple addition. Negative position values are allowed, though probably will never be generated by the layoutting algorithm.
 - Size is in absolute units and cannot be scaled. When propagating, it's copied verbatim.
 - Depth is a two part value. The upper two bytes contains the "canvas depth", which is a root level depth defined by the root node. The lower two bytes are incremented every time we go down a parent-child layer in the UI hierarchy. This enables us to have 65536 possible layers within a canvas, while having up to 65536 different layered canvases across a screen.  This is using a u32 instead of a u16 because the struct is already 4-aligned from the f32 values, might as well use those two bytes instead of letting it be padding.

Open questions:

 - Can we get away with integer values here? or do we need to use `f32`?
 - Can we SIMD accelerate this?

**Comment by alice-i-cecile** [comment]
# Coordinate traits

**Comment by alice-i-cecile** [comment]
# Orientation trait

**Comment by alice-i-cecile** [comment]
# Heritable trait

**Reply by james7132** [reply]
Assuming we have a `Heritable` trait defined as the following:

```rust
pub trait Heritable {
   type Source: Component;
   
   fn inherit(&mut self, parent: &Self, child: &Self::Source);
}

impl Heritable for GlobalTransform {
   type Source = Transform;
   ...
}

impl Heritable for ComputedVisibility {
   type Source =  Visibility;
   ...
}
```

There are a few things to note with this initial set of assumptions if we are to generalize the existing `transform_propagate_system`.

 - Root queries are no longer sufficient unless we allow skipping levels. `transform_propagate_system` relies on every Entity from the root to the leaf to have the inheritable components.
 - Allowing skipping levels requires full hierarchy traversals for all inheritance systems, not just the hierarchies with the components present, which would negatively impact performance.
 - Top down inheritance is the only mode supported. Bottoms-up composition (i.e. AABBs as a BVH) isn't well supported. This could be supported via a different trait if need be.

I definitely think we should have a trait like this if we are to add additional transform systems for 2D and UI use cases as the hierarchy traversal and inheritance patterns almost all the same across them.

**Reply by finegeometer** [reply]
I don't know what the best inheritance abstraction will be. But here's *a* useful generalization of what we do for `Transform`.

Let `T` be a monoid. In other words:
- `T` is a type.
- `T : Mul<Rhs = T, Output = T>`
- For every possible `a`, `b`, and `c` of type `T`, `a * (b * c)` and `(a * b) * c` have the same result.
- There is a designated value `I` of type `T`, called the identity.
- For every possible `a` of type `T`, the expressions `I * a`, `a`, and `a * I` have the same result.

`T` is our generalization of `Transform`. Exercise: confirm for yourself that these five conditions hold for `T = Transform`, up to floating point error.

Our generalization of `GlobalTransform` will be `struct Inherited<T>(T);`.

Our generalization of `transform_propagate_system` will be a system with the following behavior:
- For every entity `e`:
    - Get the `T` values for each ancestor of `e` (including `e` itself), in order from root to `e`.
    - Multiply the `T`s, in order. (The product of zero `T`s is defined to be the identity, `I`.)
    - Set `e`'s `Inherited<T>` to the result.

Note that this is a description of the intended *behavior*; the *implementation* could certainly be more efficient than what I described.

---

I don't suggest this as a final design, but as a starting point for discussion. In particular, I know that @cart has expressed concerns about `Inherited<T>` in a comment on #4168, but I'm having trouble understanding them.

**Reply by james7132** [reply]
That definition seems to be congruent with the overall trait design that I provided above `parent * child`  can be expressed as `child.inherit(&parent, &source)`, assuming the trait is properly implemented. Though this is built with traits and associated types instead of a generic struct `Inherited<T>`, though it wouldn't be unreasonable to build such a struct using the associated type.

If you haven't already seen it, I whipped up a preliminary implementation of this in https://github.com/bevyengine/bevy/pull/4216.
